locale_control: add locale1 dbus support

sergio-costas opened the pull request:
This PR adds support for the org.freedesktop.locale1 DBus interface. It is required for Ubuntu Core Desktop and for the installer.

Fix git checkout -b UDENG-1110-add-locale-1-dbus-interface-support-in-snapd

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

pedronis commented:
why adding these to account-control?

sergio-costas commented:
@pedronis It's just a draft, precisely to discuss where to put that. We need it in core desktop, and it seems that the installer people requires it too for classic desktop.

sergio-costas commented:
@pedronis I put it in "account_control" because it is something that has to do with the accounts, and it's required in two cases where also creating new accounts is required. But it probably deserves its own, not privileged, interface.

sergio-costas commented:
Also, we are still testing it. Maybe it will require more things...

pedronis commented:
let's see what security says, also I just noticed there's a locale-control interface (but it has some puzzling comments in it that I didn't test out)

sergio-costas commented:
Yes, but it seems that it allows to manage the global one, not the local one... although that could be a better place for this.

sergio-costas commented:
About the comments, I presume that the problem is that, in core, since /etc/default is read-only, this interface can't work.

sergio-costas commented:
@pedronis Moved into locale_control.

sergio-costas commented:
@pedronis We added this in the core-desktop snap because we needed it, but also did some little changes to make it work as expected:

- Moved _/etc/default/locale_ and _/etc/default/keyboard_ into _/etc/writable/default_
- Made two soft-links to those files in their original placement
- Created a modified systemd .deb package (where the localed daemon is located) with all the paths that pointed to _/etc/default_ renamed to _/etc/writable/default_ (because the code for both files first create a temporary file in the destination folder, and only when it is ready, it renames it to the final name, so the code must work at that folder)

This last change is quite similar to the patch that Debian/Ubuntu already do, because the default paths in the upstream systemd aren't at /etc/default.

What do you think about doing these same changes to Core? (in fact, as you see, there is a TODO in the locale-control interface that seems to point in this direction...)

jhenstridge reviewed (commented):
This looks broadly in line with the other access this interface provides:

1. a regular user can read the system locale settings via D-Bus property access
2. an administrator can change the system locale settings via the Set* D-Bus methods.

It expands the meaning of "administrator" from "running as the root account" to "can pass a polkit authorisation check", but that seems broadly okay.

We ship systemd-localed in all core* base snaps, so there's an argument for making it `implicitOnCore: true` without worrying about whether it is Core Desktop or not. The write methods may not work until we patch systemd, but we'll probably eventually have non-Desktop base snaps with the patches  the same way as for the systemd-timedated changes.

I think this is what the required support work for that would look like:

1. In the core* snap, make `/etc/locale.conf` and `/etc/vconsole.conf` symlinks to `/etc/writable`, with empty files in the initial version of `/etc/writable`.
2. Patch systemd-localed to check if these files are symlinks. If so, replace the file they point to rather than the file in /etc.

There is also code in systemd-localed to try and regenerate locales, but that appears to be disabled if the `locale-gen` executable is missing, which it is in the base snaps.

pedronis commented:
@alexmurray asking for feedback even if it is marked draft

sergio-costas commented on interfaces/builtin/locale_control.go:
```diff
@@ -33,6 +33,31 @@ const localeControlBaseDeclarationSlots = `
 const localeControlConnectedPlugAppArmor = `
 # Description: Can manage locales directly separate from 'config ubuntu-core'.
 
+# Introspection of org.freedesktop.locale1
```
Done.

sergio-costas commented on interfaces/builtin/locale_control.go:
```diff
@@ -33,6 +33,31 @@ const localeControlBaseDeclarationSlots = `
+	bus=system
+	path=/org/freedesktop/locale1
+	interface=org.freedesktop.locale1
+	member={SetLocale,SetX11Keyboard,SetVConsoleKeyboard},
```
Is it possible to have several "peer=" rules in the same entry? Or I should have to put it as

    peer=({label=unconfined,name=org.freedesktop.locale1})

or should it be

    peer=(label=unconfined, name=org.freedesktop.locale1)

?

sergio-costas replied:
About the patches for systemd-locale, I already did them for core-desktop. I'm preparing now patches for core.

sergio-costas replied:
I'll try with the second format.

sergio-costas replied:
Ok, I tried:

`peer=({label=unconfined,name=org.freedesktop.locale1})`

`peer={(label=unconfined),(name=org.freedesktop.locale1)}`

`peer=(label=unconfined, name=org.freedesktop.locale1)`

and

```
peer=(name=org.freedesktop.locale1)
peer=(label=unconfined)
```

and none of them work, so I have to duplicate the entry.

alexmurray commented on interfaces/builtin/locale_control.go:
```diff
@@ -33,6 +33,40 @@ const localeControlBaseDeclarationSlots = `
+	path=/org/freedesktop/locale1
+	interface=org.freedesktop.locale1
+	member={SetLocale,SetX11Keyboard,SetVConsoleKeyboard}
+	peer=(name=org.freedesktop.locale1),
```
Is there value in the duplicated `peer=(label=unconfined)` instance of this rule? Since we want to support both a confined and unconfined version then I think it would be easier to just drop that version and just use the `peer=(name=org.freedesktop.locale1)` rule since that should capture both the unconfined and confined cases. In that case, adding a comment like the standard `do not use peer=(label=unconfined) here since this can be DBus activated` one would be worthwhile.

sergio-costas replied:
You are right, it seems to work. Changing it.

sergio-costas replied:
Fixed.

alexmurray reviewed (approved):
LGTM - as @jhenstridge pointed out, this access is still guarded by polkit so this does not appreciably change the security semantics of this interface. One minor query re the duplicated apparmor rule.

pedronis commented on interfaces/builtin/locale_control.go:
```diff
@@ -33,6 +33,34 @@ const localeControlBaseDeclarationSlots = `
+	interface=org.freedesktop.locale1
+	member={SetLocale,SetX11Keyboard,SetVConsoleKeyboard}
+	peer=(name=org.freedesktop.locale1),
+# Receive Accounts property changed events
```
this comment seems wrong

sergio-costas replied:
Mmm... but then, it is also wrong in time_control, timeserver_control, timezone_control and upower_observe...

sergio-costas replied:
(the magic of Copy&Paste :-D )

pedronis reviewed (approved):
thanks, see remark

pedronis commented:
@jhenstridge this needs your re-review

jhenstridge reviewed (approved):
Looks good!
